### PR TITLE
For #3275: Cleans up feedback from James

### PR DESCRIPTION
**Pull request by sblatz**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/fenix/blob/master/CHANGELOG.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

**Comment by sblatz** [comment]
Going to wait on @jhugman's approval to confirm I addressed his concerns before merging 😄

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt [hunk 1]
```diff
@@ -775,6 +767,19 @@ class HomeFragment : Fragment(), CoroutineScope, AccountObserver {
         }
     }
 
+    private fun SessionManager.filteredSessions(private: Boolean, pendingSessionDeletionId: String?): List<Session> {
+        return this.sessions
+            .filter { private == it.private }
+            .filter { pendingSessionDeletionId != it.id }
+    }
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt]: Suggest:

```kotlin
private fun SessionManager.filteredSessions(private: Boolean, sessionFilter: (Session) -> Boolean): List<Session> {
    return this.sessions
        .filter { private == it.private }
        .filter(sessionFilter)
}
```

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt [hunk 2]
```diff
@@ -545,7 +547,10 @@ class HomeFragment : Fragment(), CoroutineScope, AccountObserver {
 
     private fun removeAllTabsWithUndo(isPrivate: Boolean) {
         val currentFilteredSessions =
-            context?.components?.core?.sessionManager?.sessions?.filter { it.private == isPrivate } ?: return
+            requireComponents.core.sessionManager.filteredSessions(
+                private = isPrivate,
+                pendingSessionDeletionId = pendingSessionDeletion?.sessionId
+            )
```
[the remark this answers is not in the document]
Comment (jhugman) [reply app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt]: Actually, what's the point of `currentFilteredSessions`, is it used anywhere?

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt [hunk 3]
```diff
@@ -775,6 +767,19 @@ class HomeFragment : Fragment(), CoroutineScope, AccountObserver {
...
+            .filter { private == it.private }
+            .filter { pendingSessionDeletionId != it.id }
+    }
+
+    private fun List<Session>.toTabs(): List<Tab> {
+        return this.map {
+            val selected = it == requireComponents.core.sessionManager.selectedSession
+            it.toTab(context!!, selected)
+        }
+    }
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt]: Suggest: 

```kotlin
private fun List<Session>.toTabs(): List<Tab> {
    val selected = sessionManager.selectedSession
    return this.map {  it.toTab(requireContext, it == selected) }
}
```

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt [hunk 4]
```diff
@@ -695,9 +691,7 @@ class HomeFragment : Fragment(), CoroutineScope, AccountObserver {
     private fun scrollToCollections(addedTabsSize: Int) {
         launch {
             delay(ANIM_SCROLL_DELAY)
-            val tabsSize = requireComponents.core.sessionManager.sessions.filter {
-                (activity as HomeActivity).browsingModeManager.isPrivate == it.private
-            }.size
+            val tabsSize = getListOfTabs().size
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt]: This call is a space & time complexity of `O(n)` (we create a new `Tab` object for each tab, only to throw them all away).

Suggest:

```kotlin
val tabSize = sessionManager.sessions.size
```

which may be still be `O(n)` (I haven't looked at the implementation of `sessions`), but you won't be creating any more objects.

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt [hunk 5]
```diff
@@ -707,9 +701,7 @@ class HomeFragment : Fragment(), CoroutineScope, AccountObserver {
     private fun animateTopCollection() {
         launch {
             delay(ANIM_SCROLL_DELAY)
-            val tabsSize = requireComponents.core.sessionManager.sessions.filter {
-                (activity as HomeActivity).browsingModeManager.isPrivate == it.private
-            }.size
+            val tabsSize = getListOfTabs().size
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt]: See above re:
```kotlin
val tabSize = sessionManager.sessions.size
```

If you do this trick another time, consider making it into a calculated property of `HomeFragment`.

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt [hunk 6]
```diff
@@ -594,27 +594,26 @@ class HomeFragment : Fragment(), CoroutineScope, AccountObserver {
...
-            .filter { pendingDeletionSessionId != it.id }
-            .map {
-                val selected = it == sessionManager.selectedSession
-                it.toTab(context, selected)
-            }
+        return context.components.core.sessionManager
+            .filteredSessions(
+                private = (activity as HomeActivity).browsingModeManager.isPrivate,
+                pendingSessionDeletionId = pendingSessionDeletion?.sessionId
+            ).toTabs()
```
[the remark this answers is not in the document]
Comment (jhugman) [reply app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt]: `filteredSessions` gerdsdarnit!

**Pull request merged**